clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -604,6 +597,76 @@ protected void checkLongRunningTaskState() throws InterruptedException {
     }
   }
 
+  private boolean canSwitchToLeaderTopic(PartitionConsumptionState pcs) {
+    /**
+     * Potential risk: it's possible that Kafka consumer would starve one of the partitions for a long
+     * time even though there are new messages in it, so it's possible that the old leader is still producing
+     * after waiting for 5 minutes; if it does happen, followers will detect upstream offset rewind by
+     * a different producer host name.
+     */
+    long lastTimestamp = getLastConsumedMessageTimestamp(pcs.getPartition());
+    if (LatencyUtils.getElapsedTimeInMs(lastTimestamp) <= newLeaderInactiveTime) {
+      return false;
+    }
+
+    /**
+     * For user meta stores, it requires all messages in local VT topic to be consumed before switching to leader topic,
+     * otherwise, it can get into an error replica issue when the followings happen:
+     *
+     * 1. Consumed RT data is still less than threshold thus has not been persisted to disk {@link StoreIngestionTask#shouldSyncOffset}.
+     *    After host restarts, the RT offset retrieved from disk is stale.
+     * 2. RT is truncated due to retention and has no data message in it.
+     * 3. Meta system store switches to RT so quickly that has not yet fully consumed data from local VT.
+     *
+     * When all above conditions happen together, meta system stores switch to RT, but subscribe to a stale offset and
+     * cannot consume anything from it (empty) and update, thus
+     *    RT end offset - offset in leader replica > threshold
+     * thus replica cannot become online {@link StoreIngestionTask#isReadyToServe}.
+     */
+    if (VeniceSystemStoreUtils.isMetaStore(storeName) && !isKafkaTopicPartitionFullyConsumed(pcs)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @return <code>true</code>, if kafka topic partition is fully consumed by comparing its offset against end offset;
+   *         <code>false</code>, otherwise.
+   */
+  private boolean isKafkaTopicPartitionFullyConsumed(PartitionConsumptionState pcs) {
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    long endOffset = getKafkaTopicPartitionEndOffSet(pcs);
```

Comment:
Since the `offset` is local version topic offset, the method is to check whether local version topic is fully consumed, right? If so, I have two asks:
1. Could we make the method name specific, i.e. something like isLocalVersionTopicPartitionFullyConsumed?
2. The `endOffset` should be the end offset of local version topic. Why do we need to get kafka url? Could we just use the local kafka url `localKafkaServer`?

Reply:
Makes sense. Will fix them.